tutorials/HDF5/Subset-HDF5-hyperspectral.Rmd

```diff
@@ -0,0 +1,241 @@
+
+```{r populate-group-attributes}
+
+# make a list of all groups within the full tile file
+ls=h5ls(f,all=T)
+
+#make a list of all of the names within the Coordinate_System group
+cg=unique(ls[ls$group=="/SJER/Reflectance/Metadata/Coordinate_System",]$name)
+
+for(i in 1:length(cg)){
+  print(cg[i])
+  a=h5readAttributes(f,paste0("/SJER/Reflectance/Metadata/Coordinate_System/",cg[i]))
+  d=h5read(f,paste0("/SJER/Reflectance/Metadata/Coordinate_System/",cg[i]))
+  attributes(d)=a
+  h5write(obj=d,file=nf,
+          name=paste0("/SJER/Reflectance/Metadata/Coordinate_System/",cg[i]),
+          write.attributes=T)
+}
+```
+
+## Spectral Subsetting
+
+The goal of subsetting this dataset is to substantially reduce the file size to make it faster to download and process these data. While each AOP mosaic tile is not particularly large in terms of its spatial scale (1km by 1km at 1m resolution= 1,000,000 pixels, or about half as many pixels at shown on a standard 1080p computer screen), the 426 spectral bands available result in a fairly large file size. While having hundreds of bands is important for advanced research, we can use far fewer bands for demonstartion and teaching purposes, while still getting a good sense of what these hyperspectral data can do. Therefore, we will reduce the spectral resolution of these data by selecting every fourth band in the dataset, which reduces the file size to 1/4 of the original!
+
+```{r subset-wavelengths}
+#First, we make out 'index'  a list of number that will allow us to select every fourth band
+idx <- seq(from = 1, to = 426, by = 4)
+
+# We then use this index to select particular wavelengths from the full tile using the "index=" argument
+ws <- h5read(file = f, 
+             name = "SJER/Reflectance/Metadata/Spectral_Data/Wavelength", 
+             index = list(idx)
+            )
+
+# As per above, we also need the wavelength attributes
+wa <- h5readAttributes(file = f, 
+                       name = "SJER/Reflectance/Metadata/Spectral_Data/Wavelength")
+attributes(ws) <- wa
+
+# Finally, write the subset of wavelengths and their attributes to the subset file
+h5write(obj=ws, file=nf,
+        name="SJER/Reflectance/Metadata/Spectral_Data/Wavelength",
+        write.attributes=T)
+```
+
+## Spatial Subsetting
+
+Even after spectral subsetting, our file size would be greater than 100Mb. Therefore, we will perform a spatial subsetting process as well when we read the hyperspectral imagery data. Note that the `index=` argument in `h5read` requires a `list()` object of three dimenstions for this example - in the order of band, column, row. Here, we are taking every fourth band (using our `idx` variabe), columns 501:1000 (the eastern half of the tile) and rows 1:500 (the northern half of the tile). The results in us extracting every fourth band of the northeast quadrant of the mosaic tile.
```

Comment:
1) add line breaks
2) add () to `h5read` to indicate it is a function.
3) in this case, I don't believe `list()` should be referred to as an "object" - isn't it a function here that contains the elements of interest?  
4) In my earlier suggestion about adding a section exploring the HDF5 file could also add in this description of the how you get 1:500 and 501:1000?

Reply:
1. Please see comment above re: line breaks.
2. Yes, should add `()` and made this correction.
3. I was wrong here, `list` should not have parentheses following. I believe that it is still technically an object (in computer programming nomenclature), but 'list' is pretty self explanatory, so I removed the word 'object'
4. Yes, that's a good idea. I will leave this 'conversation' open until I add that section and clarify these points.